### ISSUE TITLE
Unhandled error after `connect()` fails

### ISSUE THREAD

**Issue by erikkaplun** [wulczer/txpostgres#20]
_I suspect that my initial comments on the issue might not be valid anymore; my most recent comment below should provide a clearer and more correct description of the issue_

I have some DB wrapper code that is handling DB restarts (i.e. disconnects) by continuously trying to reestablish the connection. Everything is more or less working, with one exception.

It seems that right after the existing connection fails, and after a new attempt also fails (it takes our PostgreSQL 10-15 seconds to accepts new connections after a restart), if you don't explicitly call `db.close(); db.del` (`db` is the `ConnectionPool` instance), the Twisted reactor keeps calling `ConnectionPool.doRead()`, which in turn eventually reaches the point where `psycopg2.connection.poll()` (`self.pollable().poll()`) is called which in turn causes en error (because the connection has already failed) which is then not pushed to the `self._pollingD` `errback` as that deferred has been previously "consumed" by the previous error, thus resulting in an `Unhandled error` being printed—and there's no way to catch this error as `self._pollingD` was the only way to do it.

So it seems that `txpostgres` is not cancelling some of the recurring calls coming from the Twisted reactor.

With an explicit `db.close(); del db` right after a connection error, everything is working smoothly—but it's surprising behaviour and really counterintuitive to manually have to do `db.close(); del db` right after a failed attempt to `db.start()`.

Here's the code that I'm using (slightly simplified):

(below the code there's also the unhandled error traceback coming from the reactor, after the connection has already failed, i.e. an error has already been passed into `self._pollingD.errback` and `self._pollingD` has been set to `None`, so the next time an error occurs, it's just re-raised, and then the reactor does `print "Unhandled error"`)

``` python
class DbUtil(object):

    _db = None
    _connecting_d = None

    @inlineCallbacks
    def _ensure_connected(self, reconnect=False):
        if not self._db or reconnect:
            if self._connecting_d:
                log.msg("already connecting, blocking")
                yield self._connecting_d
                log.msg("...unblocked")
                return

            log.msg('connecting')
            self._connecting_d = Deferred()

            if reconnect:
                log.msg('closing defunct DB')
                self.close()
                yield sleep(5.0)
                log.msg('woke')

            while True:
                try:
                    log.msg('starting ConnectionPool...')
                    self._db = txpostgres.ConnectionPool('pyPgSQL.PgSQL', 'dbname=txpostgres_test')
                    yield self._db.start()
                except Exception as e:
                    log.err("error while connecting: %s" % type(e).__name__)
                    self.close()
                    yield sleep(2.0)
                    log.msg("woke")
                else:
                    log.msg("...DONE")
                    break
            d, self._connecting_d = self._connecting_d, None
            d.callback(None)

    @inlineCallbacks
    def _run_with_retry(self, f):
        while True:
            try:
                ret = yield f()
            except (DatabaseError, OperationalError) as e:
                log.err("database error %s; attempting to reconnect" % type(e).__name__)
                yield self._ensure_connected(reconnect=True)
            else:
                returnValue(ret)

    @inlineCallbacks
    def query(self, sql, *args, **kwargs):
        yield self._ensure_connected()
        ret = yield self._run_with_retry(lambda: self._db.runQuery(sql, *args, **kwargs))
        returnValue(ret)

    @inlineCallbacks
    def query_one(self, sql, *args, **kwargs):
        yield self._ensure_connected()
        res = yield self._run_with_retry(lambda: self.query(sql, *args, **kwargs))
        returnValue(res[0] if res else None)

    @inlineCallbacks
    def query_none(self, sql, *args, **kwargs):
        yield self._ensure_connected()
        yield self._run_with_retry(lambda: self._db.runOperation(sql, *args, **kwargs))

    def close(self):
        if self._db:
            db, self._db = self._db, None
            db.close()
```

The unhandled error traceback:

```
2012-02-02 17:35:12+0200 [connection] Unhandled Error
    Traceback (most recent call last):
      File "/Users/erik.allik/.virtualenvs/dts-agent/lib/python2.7/site-packages/twisted/python/log.py", line 84, in callWithLogger
        return callWithContext({"system": lp}, func, *args, **kw)
      File "/Users/erik.allik/.virtualenvs/dts-agent/lib/python2.7/site-packages/twisted/python/log.py", line 69, in callWithContext
        return context.call({ILogContext: newCtx}, func, *args, **kw)
      File "/Users/erik.allik/.virtualenvs/dts-agent/lib/python2.7/site-packages/twisted/python/context.py", line 118, in callWithContext
        return self.currentContext().callWithContext(ctx, func, *args, **kw)
      File "/Users/erik.allik/.virtualenvs/dts-agent/lib/python2.7/site-packages/twisted/python/context.py", line 81, in callWithContext
        return func(*args,**kw)
    --- <exception caught here> ---
      File "/Users/erik.allik/.virtualenvs/dts-agent/lib/python2.7/site-packages/twisted/internet/selectreactor.py", line 150, in _doReadOrWrite
        why = getattr(selectable, method)()
      File "/Users/erik.allik/.virtualenvs/dts-agent/lib/python2.7/site-packages/txpostgres/txpostgres.py", line 595, in doRead
        _PollingMixin.doRead(self)
      File "/Users/erik.allik/.virtualenvs/dts-agent/lib/python2.7/site-packages/txpostgres/txpostgres.py", line 158, in doRead
        self.continuePolling()
      File "/Users/erik.allik/.virtualenvs/dts-agent/lib/python2.7/site-packages/txpostgres/txpostgres.py", line 127, in continuePolling
        state = self.pollable().poll()
    psycopg2.OperationalError: asynchronous connection failed
```
